### PR TITLE
CUP 2

### PR DESCRIPTION
**Pull request by inebritov**
Hi all, we've done a patch for CUP2 support (required for 1.3.99 protocol).
For all signed requests it generated ETag with ecsdsa signature. Everything is done by middleware (called CUP2Middleware) and works pretty nice with upstream Omaha clients.
Please take a look!

### CODE REVIEW

File: omaha_server/omaha/request.xsd [hunk 1]
```diff
@@ -117,13 +122,17 @@
             <xs:attribute name="time_since_download_start_ms" use="optional" type="xs:integer"/>
             <xs:attribute name="nextversion" use="optional" type="xs:integer"/>
             <xs:attribute name="previousversion" use="optional" type="xs:integer"/>
+            <xs:attribute name="url" use="optional" type="xs:string"/>
+            <xs:attribute name="is_bundled" use="optional" type="xs:string"/>
```
Comment (evgsid) [inline omaha_server/omaha/request.xsd]: It should be `type="xs:integer"`, because really it's bool value.

File: omaha_server/omaha/response.xsd [hunk 2]
```diff
@@ -18,7 +18,7 @@
     <xs:element name="daystart">
         <xs:complexType>
             <xs:attribute name="elapsed_seconds" use="required" type="xs:positiveInteger"/>
-            <xs:attribute name="elapsed_days" use="optional" type="xs:nonPositiveInteger" default="0"/>
+            <xs:attribute name="elapsed_days" use="optional" type="xs:integer" default="0"/>
```
Comment (evgsid) [inline omaha_server/omaha/response.xsd]: I think, `xs:positiveInteger` should be OK.

File: omaha_server/omaha/request.xsd [hunk 3]
```diff
@@ -90,7 +91,10 @@
             <xs:attribute name='experiments' use='optional' type="xs:string"/>
             <xs:attribute name='iid' use='optional' type="GuidType"/>
             <xs:attribute name='installage' use='optional' type='xs:integer'/>
+            <xs:attribute name='installdate' use='optional' type='xs:integer'/>
             <xs:attribute name='tag' use='optional' type='xs:string'/>
+            <xs:attribute name='_numaccounts' use='optional' type='xs:string'/>
+            <xs:attribute name='_numsignedin' use='optional' type='xs:string'/>
```
Comment (evgsid) [inline omaha_server/omaha/request.xsd]: It's good idea to add additional rules:

```
<xs:attribute name='cohort' use='optional' type='xs:string'/>
<xs:attribute name='cohorthint' use='optional' type='xs:string'/>
<xs:attribute name='cohortname' use='optional' type='xs:string'/>
```
They are may appeared in client request.

**Comment by evgsid** [comment]
Is there any plans to apply this pull request?

File: omaha_server/omaha/request.xsd [hunk 4]
```diff
@@ -50,6 +50,7 @@
             <xs:attribute name='testsource' type='xs:NCName' use="optional"/>
             <xs:attribute name='dedup' type='xs:NCName' use="optional"/>
             <xs:attribute name='updaterchannel' type='xs:NCName' use="optional"/>
+            <xs:attribute name='shell_version' type='xs:string' use="optional"/>
```
Comment (evgsid) [inline omaha_server/omaha/request.xsd]: You can add also the next rules here:

```
<xs:attribute name='periodoverridesec' type='xs:integer' use="optional"/>
<xs:attribute name='dlpref' type='xs:string' use="optional"/>
```

File: omaha_server/omaha/request.xsd [hunk 5]
```diff
@@ -117,13 +122,17 @@
...
+            <xs:attribute name="url" use="optional" type="xs:string"/>
+            <xs:attribute name="is_bundled" use="optional" type="xs:string"/>
         </xs:complexType>
     </xs:element>
     <xs:element name='ping'>
         <xs:complexType>
             <xs:attribute name='active' use='optional' type='xs:integer' default="0"/>
             <xs:attribute name='r' use='optional' type='xs:integer' default="0"/>
             <xs:attribute name='a' use='optional' type='xs:integer' default="0"/>
+            <xs:attribute name='rd' use='optional' type='xs:string' default="0"/>
```
Comment (evgsid) [inline omaha_server/omaha/request.xsd]: Default value is -2 as mentioned it protocol documentation.
You can add also

```
 <xs:attribute name='ad' use='optional' type='xs:integer' default="-2"/>
```
rule.

File: omaha_server/omaha/request.xsd [hunk 6]
```diff
@@ -90,7 +91,10 @@
             <xs:attribute name='experiments' use='optional' type="xs:string"/>
             <xs:attribute name='iid' use='optional' type="GuidType"/>
             <xs:attribute name='installage' use='optional' type='xs:integer'/>
+            <xs:attribute name='installdate' use='optional' type='xs:integer'/>
             <xs:attribute name='tag' use='optional' type='xs:string'/>
+            <xs:attribute name='_numaccounts' use='optional' type='xs:string'/>
```
Comment (evgsid) [inline omaha_server/omaha/request.xsd]: `_num*` attributes should have `type='xs:integer'` instead of `string`.

**Comment by kyakovenko** [comment]
Hello @inebritov and @evgsid,

This PL looks cool but first of all we must make a code review. This process is slightly delayed because this task has the lowest priority for us. However, we're planning to check it during next several weeks and after that we'll be able to decide whether we can apply it.

In any case, we'll provide you a feedback.

P.S. Once again, thank you for your PL.

File: omaha_server/omaha/request.xsd (hunk 6 above)
Comment (evgsid) [inline omaha_server/omaha/request.xsd]: BTW, this fields is software-specific fields (in this case, chromium-specific), and can be removed.

**Comment by kyakovenko** [comment]
Hello guys,

We've checked your pull request, it looks great. But there is a problem, we can't apply it directly to the master branch. Could you resend your pull request for the dev branch in order to keep your authorship?

**Comment by inebritov** [comment]
Closed to resend for dev branch.

**Pull request closed without merging**